cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/AuthBinding.java

```diff
@@ -614,71 +620,81 @@ private SentryGenericServiceClient getClient() throws Exception {
+  private EntityType toEntityId(TAuthorizable tAuthorizable, Map<EntityType, String> entityParts) {
+    Authorizable sentryAuthorizable = ModelAuthorizables.from(tAuthorizable.getType(), tAuthorizable.getName());
     switch (Authorizable.AuthorizableType.valueOf(tAuthorizable.getType())) {
       case INSTANCE:
-        return new InstanceId(instanceName);
+        entityParts.put(EntityType.INSTANCE, instanceName);
+        return EntityType.INSTANCE;
       case NAMESPACE:
-        Namespace namespace = (Namespace) authorizable;
-        return new NamespaceId(namespace.getName());
+        Namespace namespace = (Namespace) sentryAuthorizable;
+        entityParts.put(EntityType.NAMESPACE, namespace.getName());
+        return EntityType.NAMESPACE;
       case ARTIFACT:
-        Artifact artifact = (Artifact) authorizable;
-        assertNonNullParent(parent, Authorizable.AuthorizableType.ARTIFACT);
-        // TODO: Ideally we should be constructing cdap authorizable from the tAuthorizable. CDAP authorizable does
-        // not have a concept of version so we pass a random version here just to create artifact id. The version will
-        // dropped automatically when an Authorizable will be created through it to list privilege.
-        return ((NamespaceId) parent).artifact(artifact.getName(), "1.0-SNAPSHOT");
+        Artifact artifact = (Artifact) sentryAuthorizable;
+        Preconditions.checkArgument(entityParts.containsKey(EntityType.NAMESPACE), "Artifact must belong to a " +
+          "namespace");
+        entityParts.put(EntityType.ARTIFACT, artifact.getName());
+        return EntityType.ARTIFACT;
       case APPLICATION:
-        Application application = (Application) authorizable;
-        assertNonNullParent(parent, Authorizable.AuthorizableType.APPLICATION);
-        return ((NamespaceId) parent).app(application.getName());
+        Application application = (Application) sentryAuthorizable;
+        Preconditions.checkArgument(entityParts.containsKey(EntityType.NAMESPACE), "Application must belong to a " +
+          "namespace");
+        entityParts.put(EntityType.APPLICATION, application.getName());
+        return EntityType.APPLICATION;
       case PROGRAM:
-        Program program = (Program) authorizable;
-        assertNonNullParent(parent, Authorizable.AuthorizableType.PROGRAM);
-        ApplicationId applicationId = (ApplicationId) parent;
-        return applicationId.program(program.getProgramType(), program.getProgramName());
+        Program program = (Program) sentryAuthorizable;
+        Preconditions.checkArgument(entityParts.containsKey(EntityType.APPLICATION), "Program must belong to a " +
+          "application");
+        StringBuilder builder = new StringBuilder();
+        if (program.getProgramType() != null) {
+          builder.append(program.getProgramType().getPrettyName().toLowerCase());
```

Comment:
It doesn't matter how we store in sentry as long as we can read it back and create authorizable out of it which is what happens and is being test in testValidEntities test present in AuthBindingEntityToAuthMapperTest. If we need extensive tests for it we can add it after the release.